benchmarks/polys.py

```diff
@@ -1,17 +1,137 @@
-import sympy
+from sympy import symbols, prod, prem, rem, degree, LC
+from sympy.polys import ZZ, Poly
 
 class TimePolyManyGens:
     """Time using a Poly with many generators"""
 
     params = [1, 10, 100, 500]
 
     def setup(self, n):
-        self.xs = sympy.symbols('x:{}'.format(n))
+        self.xs = symbols('x:{}'.format(n))
         self.x = self.xs[n // 2]
-        self.px = sympy.Poly(self.x, self.xs)
+        self.px = Poly(self.x, self.xs)
 
     def time_create_poly(self, n):
-        sympy.Poly(self.x, self.xs)
+        Poly(self.x, self.xs)
 
     def time_is_linear(self, n):
         self.px.is_linear
+
+
+class _LinearDenseQuadraticGCD:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = (1 + x + sum(y[:n])) ** 2
+        f = D * (-2 + x - sum(y[:n])) ** 2
+        g = D * (2 + x + sum(y[:n])) ** 2
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _SparseGCDHighDegree:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = 1 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)])
+        f = D * (-2 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)]))
+        g = D * (2 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)]))
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _QuadraticNonMonicGCD:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = 1 + x ** 2 * y[0] ** 2 + sum([y[i] ** 2 for i in range(1, n)])
+        f = D * (-1 + x ** 2 - y[0] ** 2 + sum([y[i] ** 2 for i in range(1, n)]))
+        g = D * (2 + x * y[0] + sum(y[1:n])) ** 2
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _SparseNonMonicQuadratic:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = -1 + x * prod(y[:n])
+        f = D * (3 + x * prod(y[:n]))
+        g = D * (-3 + x * prod(y[:n]))
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
```

Comment:
This function is still mixed up. It is assigning `self.x` but also returning `x`. What is the design here?

When you write a function you need to decide whether the function is supposed to have side effects (like modifying attributes of self) or whether it is supposed to return something. Most functions should just return something and should not have any side effects.

Sometimes (and this is one of those cases) it might be good to have a function that has side effects. In that situation though we should be clear that the intention of the function is to have side effects and so it should typically not return anything.

The assignment to `self.x` above does nothing useful because outside of this function nothing else ever accesses the attribute: it is assigned here and then only accessed here.